### PR TITLE
backend - get users in flat

### LINKED ISSUE #84: Backend - get users in a flat
**Is your feature request related to a problem? Please describe.**
currently we do not have a way to get all of the users in a house

**Describe the solution you'd like**
update endpoint `GET /house` so it also return `users` which is an array of UserResponseDTO
update endpoint `PUT /house` so it assigns the user to another house 

### Checklist:

- [x] I have double checked that the issue does not currently exist. 
- [x] I have added the appropriate labels (e.g. feature, front-end, back-end).
- [x] I have added the **Platform Project** project

### PR DESCRIPTION
**Pull request by kaylward**
# Description

````users```` property is now added to the ````House```` schema

```` GET /house ```` now also returns a list of UserResponseDTO

```` PUT /house ```` now assigns the user to another house

```` POST /house ```` now adds the owner to ````users```` as well

Fixes/resolves #84 

## Screenshots
Please include any screenshots or media that illustrates the changes made

## Type of change
Please delete options that are not relevant.

- [ ] **Bug fix** (non-breaking change which fixes an issue)
- [ ] **New feature** (non-breaking change which adds functionality)
- [x] **Improvement** (non-breaking change which improves existing functionality)
- [ ] **Breaking change** (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

# Checklist:
Leave blank if not applicable

I have completed these steps when making this pull request:
- [x] I have assigned my name to the issue
- [x] I have moved the issue to the **In Progress** column
- [x] I have labelled the PR appropriately
- [x] I have assigned myself to the PR

Before opening the PR for review:
- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my code
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes
- [x] I have moved the linked issue to the **Review in Progress** column

### CODE REVIEW

**Review by simoncheng987 — DISMISSED** [review]

**Review by AbhiRehal — DISMISSED** [review]
LGTM

File: packages/backend/src/controllers/house/house.controller.ts [hunk 1]
```diff
@@ -46,19 +48,30 @@ export class HouseController {
...
+    createHouseDto.owner = owner._id;
+    const userList: Array<Types.ObjectId> = [];
+    userList.push(owner._id);
     const house = await this.houseStoreService.create(createHouseDto);
     await this.userStoreService.updateByFirebaseId(user.uid, {
       house: house._id,
     });
+    await this.houseStoreService.update(house.id, {
+      users: userList,
+    });
```
Comment (mped822) [inline packages/backend/src/controllers/house/house.controller.ts]: Is there a reason why the house document is being updated immediately after it is being created? You could just add the usersList to the createHouseDto before you call create on the houseStoreService. For example, replacing lines 51 - 61 with the following:

```
const owner = await this.userStoreService.findOneByFirebaseId(user.uid);
createHouseDto.owner = owner._id;

const userList: Array<Types.ObjectId> = [];
userList.push(owner._id);
createHouseDto.users = userList;

const house = await this.houseStoreService.create(createHouseDto);
await this.userStoreService.updateByFirebaseId(user.uid, {
      house: house._id,
 });
```

File: packages/backend/src/controllers/house/house.controller.ts [hunk 2]
```diff
@@ -75,13 +88,23 @@ export class HouseController {
...
       const house = await this.houseStoreService.findOne(userDoc.house);
+      const userList: Array<UserResponseDto> = [];
+      for (const id of house.users) {
+        const user = await this.userStoreService.findOne(id);
+        const userDto = {
+          house: house.code,
+          firebaseId: user.firebaseId,
+        };
+        userList.push(userDto);
+      }
```
Comment (mped822) [inline packages/backend/src/controllers/house/house.controller.ts]: Could we use populate instead of this? This would need to be done inside the houseStoreService. 
https://mongoosejs.com/docs/populate.html#population

File: packages/backend/src/controllers/house/house.controller.ts [hunk 3]
```diff
@@ -100,22 +123,35 @@ export class HouseController {
     @Body() joinHouseDto: JoinHouseDto,
     @User() user: DecodedIdToken,
   ): Promise<HouseResponseDto> {
+    const addedUser = await this.userStoreService.findOneByFirebaseId(user.uid);
     const house = await this.houseStoreService.findOneByCode(
       joinHouseDto.houseCode,
     );
-
+    house.users.push(addedUser._id);
```
Comment (mped822) [inline packages/backend/src/controllers/house/house.controller.ts]: This should go after the null check on line 131

File: packages/backend/src/controllers/house/house.controller.ts [hunk 4]
```diff
@@ -100,22 +123,35 @@ export class HouseController {
...
-
+      const userList: Array<UserResponseDto> = [];
+      for (const id of house.users) {
+        const user = await this.userStoreService.findOne(id);
+        const userDto = {
+          house: house.code,
+          firebaseId: user.firebaseId,
+        };
+        userList.push(userDto);
+      }
```
Comment (mped822) [inline packages/backend/src/controllers/house/house.controller.ts]: Should be able to use populate instead for this too

File: packages/backend/src/controllers/house/house.controller.ts (hunk 1 above)
Comment (justcrossheaven) [reply packages/backend/src/controllers/house/house.controller.ts]: Thanks for the advice! I actually tried this before but had some undefined error messages. And I just realized I simply did not declare the array type properly... Will update

File: packages/backend/src/controllers/house/house.controller.ts (hunk 2 above)
Comment (justcrossheaven) [reply packages/backend/src/controllers/house/house.controller.ts]: In this way we would need to inject the userStoreService to the houseStoreService to get the user right? Which will eventually end up causing errors in house service test.

File: packages/backend/src/controllers/house/dto/house-response.dto.ts [hunk 5]
```diff
@@ -11,4 +12,6 @@ export default class HouseResponseDto {
   code: string;
   @ApiProperty({ required: false })
   owner: string;
+  @ApiProperty({ required: false })
+  users: UserResponseDto[];
```
Comment (j-chad) [inline packages/backend/src/controllers/house/dto/house-response.dto.ts:16]: why is this not required? There should always be at least one user

File: packages/backend/src/controllers/house/dto/house-response.dto.ts [hunk 6]
```diff
@@ -11,4 +12,6 @@ export default class HouseResponseDto {
   code: string;
   @ApiProperty({ required: false })
   owner: string;
```
Comment (j-chad) [inline packages/backend/src/controllers/house/dto/house-response.dto.ts:14]: this should be required, not sure why it wasn't already

File: packages/backend/src/controllers/house/dto/house-response.dto.ts [hunk 7]
```diff
@@ -11,4 +12,6 @@ export default class HouseResponseDto {
   code: string;
```
Comment (j-chad) [inline packages/backend/src/controllers/house/dto/house-response.dto.ts:12]: should be required, not sure why it wasn't already

File: packages/backend/src/controllers/house/house.controller.ts [hunk 8]
```diff
@@ -46,19 +48,30 @@ export class HouseController {
     @User() user: DecodedIdToken,
   ): Promise<HouseResponseDto> {
     createHouseDto.code = this.houseUtil.generateString(8);
-    createHouseDto.owner = (
-      await this.userStoreService.findOneByFirebaseId(user.uid)
-    )._id;
+    const owner = await this.userStoreService.findOneByFirebaseId(user.uid);
+    createHouseDto.owner = owner._id;
+    const userList: Array<Types.ObjectId> = [];
+    userList.push(owner._id);
```
Comment (j-chad) [inline packages/backend/src/controllers/house/house.controller.ts]:
```suggestion
    const userList: Array<Types.ObjectId> = [owner._id];
```

File: packages/backend/src/controllers/house/house.controller.ts [hunk 9]
```diff
@@ -46,19 +48,30 @@ export class HouseController {
...
     const house = await this.houseStoreService.create(createHouseDto);
```
Comment (j-chad) [inline packages/backend/src/controllers/house/house.controller.ts]: It will probably be easier to create the house like this rather than assigning to the DTO

```
    const house = await this.houseStoreService.create({
         code: 'dsfsfs',
         owner: 'sfsdff',
         users: ['asdasdsad']
    });
```

File: packages/backend/src/controllers/house/house.controller.ts (hunk 1 above)
Comment (j-chad) [reply packages/backend/src/controllers/house/house.controller.ts]: the createHouseDTO should not expose the users list. But @mped822 is correct in that you shouldn't be updating after creation. Instead have a look at my suggestion to create the house

File: packages/backend/src/controllers/house/house.controller.ts [hunk 10]
```diff
@@ -46,19 +48,30 @@ export class HouseController {
...
+    };
+    userSet.push(userDto);
     return {
       code: house.code,
       address: house.address,
       email: house.email,
       owner: user.uid,
       name: house.name,
+      users: userSet,
     };
```
Comment (j-chad) [inline packages/backend/src/controllers/house/house.controller.ts]:
```suggestion
    return {
      code: house.code,
      address: house.address,
      email: house.email,
      owner: user.uid,
      name: house.name,
      users: [
          {
              house: house.code,
              firebaseId: owner.firebaseId
          }
      ],
    };
```

File: packages/backend/src/controllers/house/house.controller.ts (hunk 2 above)
Comment (j-chad) [reply packages/backend/src/controllers/house/house.controller.ts]: you shouldn't need to inject the service. Please take a look at the docs that @mped822 linked

File: packages/backend/src/db/house/house.schema.ts [hunk 11]
```diff
@@ -19,6 +19,9 @@ export class House {
 
   @Prop({ type: Types.ObjectId, ref: 'User' })
   owner: Types.ObjectId;
+
+  @Prop()
```
Comment (j-chad) [inline packages/backend/src/db/house/house.schema.ts]: there is probably a way to tell mongoose that it is an array of ids, similar to line 20 above.

File: packages/backend/src/db/house/houseStore.service.spec.ts [hunk 12]
```diff
@@ -87,6 +88,7 @@ describe('HouseStoreService', () => {
       address: 'Mars',
       code: 'lol',
       owner: null,
+      users: null,
```
Comment (j-chad) [inline packages/backend/src/db/house/houseStore.service.spec.ts]: why null?

**Review by j-chad — DISMISSED** [review]

**Comment by j-chad** [comment]
also can you please add tests

File: packages/backend/src/controllers/house/house.controller.ts (hunk 9 above)
Comment (justcrossheaven) [reply packages/backend/src/controllers/house/house.controller.ts]: You mean sth like this?
````  
const house = await this.houseStoreService.create({
      name: createHouseDto.name,
      email: createHouseDto.email,
      address: createHouseDto.address,
      code: this.houseUtil.generateString(8),
      owner: owner._id,
      users: userList
    });

````

Comment (j-chad) [reply packages/backend/src/controllers/house/house.controller.ts]: you can just do the user list inline but yeah

File: packages/backend/src/db/house/houseStore.service.spec.ts (hunk 12 above)
Comment (justcrossheaven) [reply packages/backend/src/db/house/houseStore.service.spec.ts]: ````users```` is a list of objectID and is not specified/created. So should be set to null to make things equal.

**Comment by justcrossheaven** [comment]
> also can you please add tests

Sorry not sure about this. We would need to have corresponding users with ids to test the getUserDto. Do you want me to add a test with empty ````users```` instead?

File: packages/backend/src/controllers/house/house.controller.ts [hunk 13]
```diff
@@ -75,13 +85,15 @@ export class HouseController {
     const userDoc = await this.userStoreService.findOneByFirebaseId(user.uid);
     if (userDoc.house != undefined) {
       const house = await this.houseStoreService.findOne(userDoc.house);
+      const userList = await this.houseStoreService.getUserDto(house.id);
```
Comment (mped822) [inline packages/backend/src/controllers/house/house.controller.ts:88]: Just create a new method in `houseStoreService` called `findOneWithUserPopulation` and call this instead on line 87. The method should contain the following: ` return await this.HouseModel.findOne({ _id: id }).populate('users').exec();`. Then there is no need for the `getUserDto` method.

File: packages/backend/src/controllers/house/house.controller.ts [hunk 14]
```diff
@@ -100,22 +112,27 @@ export class HouseController {
...
+      house.users.push(addedUser._id);
       await this.userStoreService.updateByFirebaseId(user.uid, {
         house: house._id,
       });
+      await this.houseStoreService.update(house.id, {
+        users: house.users,
+      });
       const owner = await this.userStoreService.findOne(house.owner);
-
+      const userList = await this.houseStoreService.getUserDto(house.id);
```
Comment (mped822) [inline packages/backend/src/controllers/house/house.controller.ts:128]: This should be changed according to the above comment

Comment (mped822) [reply packages/backend/src/controllers/house/house.controller.ts:128]: If you get the house with population on line 116 then you can just set the users on line 135 to `house.users`

Comment (justcrossheaven) [reply packages/backend/src/controllers/house/house.controller.ts:128]: Currently the UserResponseDto is not extending UserModel or UserDocument, so I think you can't just set the users to be populated house.users? 

Also the house on line116 is found by houseCode while the ````GET /house```` is thru Id, so unless you are suggesting the following way:
````
return await this.HouseModel.findOne({$or: [{ _id: id }, {code: code}]}).populate('users').exec();
````
with optional parameters, I think we would need to create two functions here.

File: .eslintcache [hunk 15]
```diff
@@ -0,0 +1 @@
...backend/src/db/house/houseStore.service.ts":"2","/home/cjia881/701-project/new-add-user/flatshare/packages/backend/src/controllers/house/dto/create-house.dto.ts":"3"},{"size":4874,"mtime":1647378354223,"results":"4","hashOfConfig":"5"},{"size":1880,"mtime":1647378396363,"results":"6","hashOfConfig":"5"},{"size":376,"mtime":1647378354223,"results":"7","hashOfConfig":"5"},{"filePath":"8","messages":"9","suppressedMessages":"10","errorCount":0,"fatalErrorCount":0,"warningCount":0,"fixableErrorCount":0,"fixableWarningCount":0},"15qdnne",{"filePath":"11","messages":"12","suppressedMessages":"13","errorCount":0,"fatalErrorCount":0,"warningCount":0,"fixableErrorCount":0,"fixableWarningCount":0},{"filePath":"14","messages":"15","suppressedMessages":"16","errorCount":0,"fatalErrorCount":0,"warningCount":0,"fixableErrorCount":0,"fixableWarningCount":0},"/home/cjia881/701-project/new-add-user/flatshare/packages/backend/src/controllers/house/house.controller.ts",[],[],"/home/cjia881/701-project/new-add-user/flatshare/packages/backend/src/db/house/houseStore.service.ts",[],[],"/home/cjia881/701-project/new-add-user/flatshare/packages/backend/src/controllers/house/dto/create-house.dto.ts",[],[]]
```
Comment (frasermcc9) [inline .eslintcache]: Remove this file and add to gitignore

File: packages/backend/src/controllers/house/house.controller.ts (hunk 14 above)
Comment (mped822) [reply packages/backend/src/controllers/house/house.controller.ts:128]: I don't think it will be an issue if you make the UserResponseDto extend the UserModel. I would try it anyway before you do this. The fields are the same so it might not be an issue (I am not a typescript expert). Just play around with it a bit and see what works. 

Also I meant to say create two functions, one for find by code and one for find by id. That is also fine in my opinion because this will improve the code and reduce the complexity by using populate instead of the loops.

Comment (justcrossheaven) [reply packages/backend/src/controllers/house/house.controller.ts:128]: I just tried it out and the house in UserReponseDto is a string representing the houseCode, and the house in UserModel is an ObjectID reference to the House.

Comment (mped822) [reply packages/backend/src/controllers/house/house.controller.ts:128]: Can you make the house in UserResponseDto an ObjectId?

Comment (justcrossheaven) [reply packages/backend/src/controllers/house/house.controller.ts:128]: I can but that would also require changing multiple previous implementations. @j-chad Any thoughts?

Comment (j-chad) [reply packages/backend/src/controllers/house/house.controller.ts:128]: I would prefer if the api returns the code. The internal ID is not useful to the front-end

File: packages/backend/src/db/house/houseStore.service.spec.ts (hunk 12 above)
Comment (j-chad) [reply packages/backend/src/db/house/houseStore.service.spec.ts]: shouldn't it just be an empty list?

File: packages/backend/src/db/house/houseStore.service.ts [hunk 16]
```diff
@@ -29,6 +30,23 @@ export class HouseStoreService {
...
+      .exec();
+    const userList: Array<UserResponseDto> = [];
+    for (const user of house.users) {
+      const userDto = {
+        house: house.code,
+        firebaseId: user['firebaseId'],
+      };
+      userList.push(userDto);
+    }
+    return userList;
```
Comment (mped822) [inline packages/backend/src/db/house/houseStore.service.ts:47]: Can you just return the `house.users`? If not, what is there is okay, but it would be simpler to return `house.users`.

File: packages/backend/src/controllers/house/dto/create-house.dto.ts [hunk 17]
```diff
@@ -10,4 +11,7 @@ export class CreateHouseDto extends HouseModel {
 
   @ApiProperty()
   readonly address: string;
+
+  @ApiProperty()
+  users: Array<Types.ObjectId>;
```
Comment (mped822) [inline packages/backend/src/controllers/house/dto/create-house.dto.ts]: This is not needed in the create house dto if we are only adding the requesting user as a user of the house in the create house endpoint. Otherwise, you need to account for users being present in this array when a house is created.

**Review by mped822 — CHANGES_REQUESTED** [review]
Can you please remove the users array from the create house dto and update any cascading issues caused by this.

File: packages/backend/.eslintrc.js [hunk 18]
```diff
@@ -7,6 +7,7 @@ module.exports = {
   },
   plugins: ['@typescript-eslint/eslint-plugin'],
   extends: [
+    'eslint:recommended',
```
Comment (ljey973) [inline packages/backend/.eslintrc.js]: Is this change necessary? The other two seem to be sufficient guidelines seem to be sufficient.

Comment (justcrossheaven) [reply packages/backend/.eslintrc.js]: This change is made based on another pr made by @j-chad . Commenting this out seems not creating issues.

File: packages/backend/src/db/house/houseStore.service.spec.ts (hunk 12 above)
Comment (justcrossheaven) [reply packages/backend/src/db/house/houseStore.service.spec.ts]: Sorry didn't notice this reply yesterday. Empty list would work as well.

File: packages/backend/src/guards/firebase.auth.ts [hunk 19]
```diff
@@ -7,7 +7,9 @@ import {
   ServiceAccount,
   Credential,
 } from 'firebase-admin/app';
-import { getAuth } from 'firebase-admin/auth';
+import { DecodedIdToken, getAuth } from 'firebase-admin/auth';
+import { UserStoreService } from 'src/db/user/userStore.service';
+import { UserModel } from 'src/db/user/user.schema';
```
Comment (mped822) [inline packages/backend/src/guards/firebase.auth.ts]: Can you make these paths relative

**Review by mped822 — APPROVED** [review]

**Review by frasermcc9 — APPROVED** [review]
LGTM
